File: sliding-window/longest-substring.js
// state, start, longest
//
// loop end
//      update stte with new end
//
//      while not valid
//          update start state
//          increment start
//
//      update max
//
//  return state

function longestSubstrintWithoutRepeat(str) {
    let start = 0;
    let charCount = {}; // # times char appears in window
    let max = 0;

    for (let end = 0; end < str.length; end++) {
        charCount[str[end]] = (charCount[str[end]] ?? 0) + 1; // update the charCount
        
        // if not valid: move up start and handle its state
        while (charCount[str[end]] > 1) {
            // update start's state:
            charCount[str[start]]--;
            // move up start
            start++;
        }

        max = Math.max(max, (end - start) + 1);
    }

    return max;
}

console.log(longestSubstrintWithoutRepeat('eghghhgg'))

File: sliding-window/longest_substring.py
def longest_substring_without_repeating_chars(str):
    start = 0
    max_count = 0
    state = {}

    for end in range(len(str)):
        state[str[end]] = state.get(str[end], 0) + 1

        while state.get(str[end], 0) > 1:
            state[str[end]] -= 1
            start += 1

        max_count = max(max_count, end - start + 1)

    return max_count


res = longest_substring_without_repeating_chars("substring")
